aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,207 @@
+    // We will advertise to the community that max expiration time for orderless txns is 60 seconds.
+    // Adding a 5 second slack here as the client's time and the blockchain's time may drift.
+    const MAX_EXPIRATION_TIME: u64 = 65;
+
+    const E_NONCE_HISTORY_DOES_NOT_EXIST: u64 = 1001;
+
+    struct Map has store {
+        // Maps NonceKey = (address, nonce) to txn expiration time.
+        nonces: BigOrderedMap<NonceKey, u64>,
+        // last_stored_time stores the timestamp at which a nonce entry is last added to the map.
+        last_stored_time: u64,
+    }
+
+    // The bucket stores (address, nonce) -> txn expiration time mappings.
+    // All the entries in the bucket contain the same hash(address, nonce) % NUM_BUCKETS.
+    // Using multiple maps to facilitate easier garbage collection of expired entries.
+    struct Bucket has store {
+        nonce_maps: vector<Map>,
+    }
+
+    struct NonceKey has copy, drop, store {
+        sender_address: address,
+        nonce: u64,
+    }
+
+    // An orderless transaction is a transaction that doesn't have a sequence number.
+    // Orderless transactions instead contain a nonce to prevent replay attacks.
+    // If the incoming transaction has the same (address, nonce) pair as a previous unexpired transaction, it is rejected.
+    // The nonce history is used to store the list of (address, nonce, txn expiration time) values of all unexpired transactions.
+    // The nonce history is used in the transaction validation process to check if the incoming transaction is valid.
+    struct NonceHistory has key {
+        // Key = sip_hash(NonceKey) % NUM_BUCKETS
+        // Value = Bucket
+        nonce_table: Table<u64, Bucket>,
+        // Used to facilitate prefill the nonce_table with empty buckets
+        // one by one using `add_nonce_bucket` method.
+        // This is the next_key to prefill with an empty bucket
+        next_key: u64,
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        initialize_nonce_table(aptos_framework);
+    }
+
+    public entry fun initialize_nonce_table(aptos_framework: &signer) {
+        if (!exists<NonceHistory>(@aptos_framework)) {
+            let table = table::new();
+            let nonce_history = NonceHistory {
+                nonce_table: table,
+                next_key: 0,
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    fun empty_bucket(): Bucket {
+        let bucket = Bucket {
+            nonce_maps: vector[
+                Map {
+                    nonces: big_ordered_map::new_with_config(0, 0, true),
+                    last_stored_time: 0,
+                },
+                Map {
+                    nonces: big_ordered_map::new_with_config(0, 0, true),
+                    last_stored_time: 0,
+                },
+            ]
+        };
+        bucket.nonce_maps[0].nonces.allocate_spare_slots(5);
```

Comment:
why only 5, I expect every bucket would have more entries in steady state? also how does this change impact performance/gas fee?

Reply:
I don't mind increasing this value. Here, we are pre-allocating 5 nodes. Each node contains an ordered map of (address, nonce) -> expiration time inside it. And we have 50k buckets. Unless we are getting an extremely high throughput, or an attacker is crafting nonces to fall into the same buckets, I don't see why we might need more than 5 storage slots.

Reply:
I actually don't expect us needing to pre-allocate anything - single slot gets automatically allocated to start with - and only after enough entries are added that single slot overflows, do we actually use second

Reply:
I'm not sure what's the expected gas behavior here, I'd expect we pre-allocate everything so the gas is predictable to users. it feels awkward some user needs to spend extra for slots and get random refund from unrelated txn in a min or so

Reply:
`nonce + expiration time + address` = 48 bytes. Assuming a slot can contain 4KB, we can store 80+ entries in one slot. 5 storage slots mean storage space for almost 400+ entries is allocated. With 50k buckets and a short expiration time, we have plenty of storage allocated.